lib/promscrape/discovery/docker/container.go

```diff
@@ -67,14 +69,51 @@ func parseContainers(data []byte) ([]container, error) {
 	return containers, nil
 }
 
-func addContainersLabels(containers []container, networkLabels map[string]*promutils.Labels, defaultPort int, hostNetworkingHost string) []*promutils.Labels {
+func addContainersLabels(containers []container, networkLabels map[string]*promutils.Labels, defaultPort int, hostNetworkingHost string, matchFirstNetwork bool) []*promutils.Labels {
+	allContainers := make(map[string]container)
+	for _, c := range containers {
+		allContainers[c.ID] = c
+	}
+
 	var ms []*promutils.Labels
 	for i := range containers {
 		c := &containers[i]
 		if len(c.Names) == 0 {
 			continue
 		}
-		for _, n := range c.NetworkSettings.Networks {
+
+		networks := c.NetworkSettings.Networks
+		networkMode := c.HostConfig.NetworkMode
+		if len(networks) == 0 {
```

Comment:
I think we should put some love to this code. Currently, it looks fragile. I don't really like this endless for loop.

 First of all, I'd suggest to keep previous behavior and alternate it under this condition:

```
if matchFirstNetwork && len(networks) == 0
  // Try to lookup shared networks
}
```